rs-matter/src/data_model/core.rs

```diff
@@ -15,142 +15,759 @@
+        let req = TimedReq::from_tlv(&get_root_node_struct(exchange.rx()?.payload())?)?;
+        debug!("IM: Timed request: {:?}", req);
+
+        let timeout_instant = req.timeout_instant(exchange.matter().epoch);
+
+        Self::send_status(exchange, IMStatusCode::Success).await?;
+
+        Ok(timeout_instant)
+    }
+
+    async fn timed_out(
+        &self,
+        exchange: &mut Exchange<'_>,
+        timeout_instant: Option<Duration>,
+        timed_req: bool,
+    ) -> Result<bool, Error> {
+        let status = {
+            if timed_req != timeout_instant.is_some() {
+                Some(IMStatusCode::TimedRequestMisMatch)
+            } else if timeout_instant
+                .map(|timeout_instant| (exchange.matter().epoch)() > timeout_instant)
+                .unwrap_or(false)
+            {
+                Some(IMStatusCode::Timeout)
+            } else {
+                None
+            }
+        };
+
+        if let Some(status) = status {
+            Self::send_status(exchange, status).await?;
+
+            Ok(true)
+        } else {
+            Ok(false)
+        }
+    }
+
+    async fn report_data(
+        &self,
+        id: u32,
+        node_id: u64,
+        rx: &[u8],
+        tx: &mut [u8],
+        exchange: &mut Exchange<'_>,
+    ) -> Result<bool, Error>
+    where
+        T: DataModelHandler,
+    {
+        let mut wb = WriteBuf::new(tx);
+
+        let req = SubscribeReq::from_tlv(&get_root_node_struct(rx)?)?;
+        let req = ReportDataReq::Subscribe(&req);
+
+        let metadata = self.handler.lock().await;
+
+        let accessor = exchange.accessor()?;
+
+        {
+            let node = metadata.node();
+            let mut attrs = node.read(&req, None, &accessor).peekable();
 
-                    driver.complete(req).await?;
+            loop {
+                let more_chunks = req
+                    .respond(&self.handler, Some(id), &mut attrs, &mut wb, false)
+                    .await?;
+
+                exchange.send(OpCode::ReportData, wb.as_slice()).await?;
+
+                if !Self::recv_status(exchange).await? {
+                    info!("Subscription {node_id:x}::{id} removed during reporting");
+                    return Ok(false);
+                }
+
+                if !more_chunks {
+                    break;
                 }
             }
         }
 
+        Ok(true)
+    }
+
+    async fn rx_buffer(&self, exchange: &mut Exchange<'_>) -> Result<Option<B::Buffer<'a>>, Error> {
+        let buffer = self.tx_buffer(exchange).await?;
+
+        if let Some(mut buffer) = buffer {
+            let rx = exchange.rx()?;
+
+            buffer.clear();
+            buffer.extend_from_slice(rx.payload()).unwrap();
+
+            exchange.rx_done()?;
+
+            Ok(Some(buffer))
+        } else {
+            Ok(None)
+        }
+    }
+
+    async fn tx_buffer(&self, exchange: &mut Exchange<'_>) -> Result<Option<B::Buffer<'a>>, Error> {
```

Comment:
This may need some doccomments (throughout  the PR seems somewhat light on these). In particular this has a sideffect of `Self::send_status` which may not be as obvious based on the name (it seems to me like I would expect tx_buffer to grab the transmit buffer). 

To keep transmission at the same level, would it maybe make sense to do the send_status in the caller of this (since the caller is the one that does other sends?)

Reply:
I'm against moving the `send_status` call to the caller, as that would mean blowing up the codebase with repetitions, as we _always_ want to send status code "Busy" when we are low on buffers.

Changing the names is a good idea. Any suggestions?

Reply:
I have trouble with the naming as the entire code path is somwhat difficult for me regarding rx/tx.

In particular I see:
  - self.buffers is a generic "buffer pool" 
  - tx_buffer just gets a buffer from this buffer pool (so nothing specific to TX yet) and sets it to a TX_BUF_SIZE size (this is tx-specific)
  - rx_buffer gets a tx_buffer (this confuses me ... why? I guess it may be some form reuse, but naming makes it ugly ... why does a rx buffer get a tx buffer and handle all the associated sideffects?) I would love some comments: is this to have a locking protocol (you always grab tx before rx to avoid deadlocks)?
  - Exchage::rx is a wrapper around a match of its rx member (should probably have been `self.rx.as_ref().ok_or(ErrorCode::InvalidState.into())` instead of the let/else it currently is for shorter code). Without documentation I actually have a hard time following what Exchange::rx is supposed to contain and what its lifetime is.
  
While reviewing this code, I find it that we lack some documentation. I seem to periodically want to comment (or even comment) on some things and then when trying out finding out that the code was actually ok and correct, it was just hard to figure out why.

For concrete suggestions (which may be wrong in part or fully):
  - exchange::rx needs some documentation/explanation
  - a separate (maybe private) method could exist of `grab_buffer` that is not rx/tx related, however it does send the busy reply. maybe `get_buffer_for_exchange` with comment that it sends busy on error.
  - tx makes use of this and resizes the buffer
  - rx makes use of this and drops it to the floor (but again .. why...how does this work if the buffer pool has more than 1 buffer available) and returns exchange::rx

Reply:
> * self.buffers is a generic "buffer pool"

True.

> * tx_buffer just gets a buffer from this buffer pool (so nothing specific to TX yet) and sets it to a TX_BUF_SIZE size (this is tx-specific)

True.

> * rx_buffer gets a tx_buffer (this confuses me ... why? I guess it may be some form reuse, but naming makes it ugly ... why does a rx buffer get a tx buffer and handle all the associated sideffects?) I would love some comments: is this to have a locking protocol (you always grab tx before rx to avoid deadlocks)?

No it is _not_ related to locking at all, it is just code reuse. I've introduced an intermediate method to make the code reuse clearer (see below).

W.r.t. locking: The generic buffer pool does not differentiate between RX and TX buffers. Hence the buffer pool simply **cannot** wait forever in the case it has no available buffers as this would deadlock. It needs to give up after some (small amount) of time waiting. The default implementation actually does not even `await` and gives up immediately.

The contract for the buffer pool is slightly more generic and includes the possibility of the pool to `await` only so that if we want to make a cleverer buffer allocation strategy that does await a little bit (instead of giving up on the transaction) we can do that.

> * Exchage::rx is a wrapper around a match of its rx member (should probably have been `self.rx.as_ref().ok_or(ErrorCode::InvalidState.into())` instead of the let/else it currently is for shorter code). Without documentation I actually have a hard time following what Exchange::rx is supposed to contain and what its lifetime is.

Right, I've put documentation on the `rx` method. It is really just a slight optimization / alternative of `recv_fetch`.

> While reviewing this code, I find it that we lack some documentation. I seem to periodically want to comment (or even comment) on some things and then when trying out finding out that the code was actually ok and correct, it was just hard to figure out why.

The API of `Exchange` at least was supposed to be completely documented. `rx` and `rx_done` comments simply got lost when I re-arranged the code during the last round-trip of changes. Sorry for that.

> 
> For concrete suggestions (which may be wrong in part or fully):
> 
> * exchange::rx needs some documentation/explanation

Done.

> * a separate (maybe private) method could exist of `grab_buffer` that is not rx/tx related, however it does send the busy reply. maybe `get_buffer_for_exchange` with comment that it sends busy on error.

Done. But I named it simply `buffer`, as it looks to me more symmetric with `tx_buffer` and `rx_buffer`. Not sure that putting a verb in the name of `buffer` and its tx/rx siblings would help. Or is there such a code style guide for rust, similar to e.g. Java, that methods should always be verbs?

> * tx makes use of this and resizes the buffer

Exactly.

> * rx makes use of this and drops it to the floor (but again .. why...how does this work if the buffer pool has more than 1 buffer available) and returns exchange::rx

`rx_buffer` also make use of `buffer` now. It does not drop anything to the floor though. It just **copies**  the RX message from the internal buffer in `Exchange` into its own buffer returned by `buffer`. This is necessary, because the interaction code had detected that it will call a handler that awaits and/or the returned interaction response is chunked, hence it cannot just block the transport RX buffer.

`rx_done` is also documented now, which would hopefully bring some extra clarity as to what exactly is dropped to the floor.